src/realm/util/file.cpp

```diff
@@ -680,20 +680,50 @@ void File::resize(SizeType size)
 }
 
 
-void File::prealloc(SizeType offset, size_t size)
+void File::prealloc(size_t size)
 {
     REALM_ASSERT_RELEASE(is_attached());
 
 #if defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L // POSIX.1-2001 version
 
-    prealloc_if_supported(offset, size);
+    prealloc_if_supported(0, size);
 
 #else // Non-atomic fallback
 
-    if (int_add_with_overflow_detect(offset, size))
-        throw std::runtime_error("File size overflow");
-    if (get_size() < offset)
-        resize(offset);
+#if REALM_IOS
+    fstore_t store = { F_ALLOCATEALL, F_PEOFPOSMODE, 0, static_cast<off_t>(size), 0 };
+    int ret = fcntl(m_fd, F_PREALLOCATE, &store);
+    if (ret == -1) {
+        int err = errno;
+        std::string msg = get_errno_msg("fcntl() inside prealloc() failed: ", err);
+    }
+    ret = ftruncate(m_fd, size);
+    if (ret != 0) {
+        int err = errno;
+        std::string msg = get_errno_msg("ftruncate() inside prealloc() failed: ", err);
+    }
+#elif REALM_ANDROID || defined(_WIN32)
+    if (size > get_size()) {
+        size_t new_size = size;
+        if (m_encryption_key) {
+            new_size = static_cast<size_t>(data_size_to_encrypted_size(size));
+            if (new_size < size) {
+                throw std::runtime_error("File size overflow");
+            }
+        }
+
+        constexpr size_t chunk_size = 4096;
+        int64_t original_size = get_size_static(m_fd);
+        seek(original_size);
+        size_t num_bytes = size_t(new_size - original_size);
+        std::string zeros(chunk_size, '\0');

```

Comment:
Do you think it's worth making this static so that we don't have to make the 4k allocation on every call to prealloc?

Reply:
Not worth it; the time it takes to write the data blocks to disk is infinitely longer than the alloc. And downside is a thread shared static with a c_str() call...

Reply:
This is probably a pragmatic solution, but it still feels kind of ugly. Isn't there an equivalent to `memset()` rather than `memcpy()`?